### PR TITLE
Filter nulls from joins where possible to improve performance.

### LINKED ISSUE #569: [BUG] left_semi_join operation is abnormal and serious time-consuming
**Describe the bug**
I test tpc-ds query-10 with rapids-0.2 recently committed , and I find there is a phenomenon that during the reduce process after shuffle, several takes are time-consuming, and I use nsys found that time spent on left_semi_join , i suspect this is a bug
Normally, this query can be completed in dozens of seconds, but now the several takes 5-8 minutes. At the same time, it is found that the GPU utilization rate is very high in this process  
image
image

### PR DESCRIPTION
**Pull request by revans2**
This fixes #569 

Although it really is just a work around for it until https://github.com/rapidsai/cudf/issues/6052 can go in.

This is caused by #349 

Essentially the performance problem showed up after we started treating nulls properly when doing a join.

### CODE REVIEW

File: integration_tests/src/main/python/tpcds_test.py [hunk 1]
```diff
@@ -35,5 +35,17 @@
 @allow_non_gpu(any=True)
 @pytest.mark.parametrize('query', queries)
 def test_tpcds(tpcds, query):
+  assert_gpu_and_cpu_are_equal_collect(
```
Comment (revans2) [inline integration_tests/src/main/python/tpcds_test.py:38]: Yes this change is some what unrelated and I can pull it out if people want me to.  I wanted to test all of the queries at a high scale factor to be sure that everything worked properly and I threw this change it with it because there were a lot of queries that were slow because of this.

**Comment by revans2** [comment]
build

**Comment by tgravescs** [comment]
I'm curious, what is the performance with the change, I assume its still a bit slower when there are nulls due to the extra filter?

These are the same rules spark uses?

**Comment by revans2** [comment]
@tgraves In the case of tpcds query 4 at a scale factor of 200 reading from parquet the worst case for this is a single join with all of the data in it.  In that case it is joining 23M compressed data in the left table and 51MB compressed data in the right table with 500,000 nulls in it (about 2% of the total right table).  Without this fix the join takes about 50 seconds. With this fix it is sub-second, and the total query finishes in about 6 to 7 seconds on a single GPU.

I have not seen any real performance degradation for joins with nulls in them, but cudf in 0.16 should be able to get a fix in for this so we can remove the extra null filtering all together, and they will do the null filtering as they build the table.  They already have this for hash aggregate and it should be simple to bring it over to the join tables as well.

File: shims/spark300/src/main/scala/com/nvidia/spark/rapids/shims/spark300/GpuHashJoin.scala [hunk 2]
```diff
@@ -110,6 +116,58 @@ trait GpuHashJoin extends GpuExec with HashJoin {
...
+      case (RightOuter, BuildLeft) => builtAnyNullable
+      case (LeftOuter | LeftAnti, BuildRight) => builtAnyNullable
+      case _ => false
+    }
+  }
+
+  private[this] lazy val shouldFilterStreamTableForNulls: Boolean = {
+    val streamedAnyNullable = gpuStreamedKeys.exists(_.nullable)
+    (joinType, buildSide) match {
+      case (_: InnerLike | LeftSemi, _) => streamedAnyNullable
```
Comment (jlowe) [inline shims/spark300/src/main/scala/com/nvidia/spark/rapids/shims/spark300/GpuHashJoin.scala:143]: Why do we need to filter the streamed keys in this case?  It seems like this would be purely slower than letting the hash join try to match the keys and fail to match.  IIUC it's the long collision chains in the built table that's the performance problem, and that's solved by eliminating the nulls in the built table (which we want to do regardless of join type).

**Comment by jlowe** [comment]
Was the omission of changes to the spark301 and spark310 shims intentional?

**Comment by revans2** [comment]
build

**Comment by jlowe** [comment]
What about the join implementations in the [spark300db](https://github.com/NVIDIA/spark-rapids/blob/branch-0.2/shims/spark300db/src/main/scala/com/nvidia/spark/rapids/shims/spark300db/GpuHashJoin.scala#L137-L140) and [spark310db](https://github.com/NVIDIA/spark-rapids/blob/branch-0.2/shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/GpuHashJoin.scala#L137-L140) shims?

Minimally I think the spark300db shim needs to be addressed.  I could see avoiding the spark310 shim for now given we don't support spark310 in the 0.2 release and expect this to be fixed in libcudf by the 0.3 release.  If so, should file a blocker for the 0.3 release to make sure it's either fixed in libcudf or update the spark310 shim accordingly.

**Comment by revans2** [comment]
You are right.  They never show up because IntelliJ does not see it. Development for them is always difficult. I'll try to refactor things to make them as small as possible.

File: shims/spark301/src/main/scala/com/nvidia/spark/rapids/shims/spark301/GpuBroadcastHashJoinExec.scala [hunk 3]
```diff
@@ -143,10 +143,15 @@ case class GpuBroadcastHashJoinExec(
     val boundCondition = condition.map(GpuBindReferences.bindReference(_, output))
 
     lazy val builtTable = {
-      // TODO clean up intermediate results...
-      val keys = GpuProjectExec.project(broadcastRelation.value.batch, gpuBuildKeys)
-      val combined = combine(keys, broadcastRelation.value.batch)
-      val ret = GpuColumnVector.from(combined)
+      val ret = withResource(
+        GpuProjectExec.project(broadcastRelation.value.batch, gpuBuildKeys)) { keys =>
+        val combined = GpuHashJoin.incRefCount(combine(keys, broadcastRelation.value.batch))
```
Comment (jlowe) [inline shims/spark301/src/main/scala/com/nvidia/spark/rapids/shims/spark301/GpuBroadcastHashJoinExec.scala:148]: Shouldn't incrementing the refcount be an implementation detail of `filterBuiltTableIfNeeded` and only if the table wasn't filtered?  Otherwise it seems like the refcounts get off if it was filtered.

**Comment by jlowe** [comment]
> They never show up because IntelliJ does not see it.

Weird, they show up for me, at least the 300, 301 and 310 versions of them.  The spark300db version doesn't, probably because I don't have the databricks profile enabled, as it won't build outside of a databricks node.

File: shims/spark301/src/main/scala/com/nvidia/spark/rapids/shims/spark301/GpuBroadcastHashJoinExec.scala (hunk 3 above)
Comment (revans2) [reply shims/spark301/src/main/scala/com/nvidia/spark/rapids/shims/spark301/GpuBroadcastHashJoinExec.scala:148]: No, because the contract with `filterBuiltTableIfNeeded` is that it will close the input table and return an output table that the caller must take ownership of and close as needed.  `combine` does not inc any reference counts when it creates `combined` so if we close it then we end up double closing things. 

In the particular case we have to leak `broadcastRelation.value.batch` because of how broadcast works.  So the best I can do is what I am doing now with closing `keys` on the off chance that they are not just a `BoundReference`, but might actually do some computation and produce a result that we were leaking before, which is what the TODO was.

Comment (revans2) [reply shims/spark301/src/main/scala/com/nvidia/spark/rapids/shims/spark301/GpuBroadcastHashJoinExec.scala:148]: I forgot to add that if you want me to add in comments to describe this I can.

**Comment by revans2** [comment]
We really need some kind of a follow on for the join shims, because the fact that I have to make the exact same change to 3x the number of files and hope that I didn't forget one, especially because I cannot compile at least one set of those changes is a problem.

**Comment by revans2** [comment]
build

**Comment by tgravescs** [comment]
we have an issue for join follow up just haven't had time yet. https://github.com/NVIDIA/spark-rapids/issues/412

File: shims/spark301/src/main/scala/com/nvidia/spark/rapids/shims/spark301/GpuBroadcastHashJoinExec.scala (hunk 3 above)
Comment (revans2) [reply shims/spark301/src/main/scala/com/nvidia/spark/rapids/shims/spark301/GpuBroadcastHashJoinExec.scala:148]: I am struggling with why this is so hard to follow.  I added comments below for `ShuffledHashJoin` which hopefully explains what is happening.

```scala
        val builtTable = withResource(ConcatAndConsumeAll.getSingleBatchWithVerification(
          buildIter, localBuildOutput)) { buildBatch: ColumnarBatch =>
          // builtBatch owns some data and will be closed when it goes out of scope.
          withResource(GpuProjectExec.project(buildBatch, gpuBuildKeys)) { keys =>
            // keys conceptually owns some data (although it could have just increased the reference counts for
            // things in builtBatch) and will be closed when it goes out of scope
            val combined = GpuHashJoin.incRefCount(combine(keys, buildBatch))
            // We call incRefCount here because conceptually we want combine to look like it copied its
            // input data, but in reality it just placed columns in a new ColumnarBatch.  We inc the ref counts here
            // So that we can close combined.
            // Conceptually combined will be closed by filterBuiltTableIfNeeded and the returned ColumnarBatch (filtered)
            // will be owned by us and need to be closed.  The reality might be a bit more complex than that.
            // Perhaps it should be called filterBuiltTableIfNeededAndCloseInput
            val filtered = filterBuiltTableIfNeeded(combined)
            combinedSize =
                GpuColumnVector.extractColumns(filtered)
                    .map(_.getBase.getDeviceMemorySize).sum.toInt
            // filtered is then closed after it is transitioned to a Table which increments the reference counts for
            // all of the columns
            withResource(filtered) { filtered =>
              GpuColumnVector.from(filtered)
            }
          }
        }
```

The broadcast code looks exactly the same as the above code.  The only difference is that there is no outer `withResources` to close `builtBatch`, because we want to leak the equivalent of `builtbatch`, and there is no `combinedSize` metric because `GpuBroadcastExchange` has a size metric in it. If you want me to make all of that code common I can do it if it will help.

Comment (jlowe) [reply shims/spark301/src/main/scala/com/nvidia/spark/rapids/shims/spark301/GpuBroadcastHashJoinExec.scala:148]: > I am struggling with why this is so hard to follow.

It's not you, it's me.  `incRefCount` was added as part of this, and it's weird to me that we've never come across a use-case that needed it before now.  That made me wonder if the refcounting was really correct here and second-guess it.  That, combined with the fact that I didn't fully understand the semantics of `combine` (no pun intended) is what tripped me up.  Apologies for the noise.
